### ISSUE TITLE
Cannot connect to Kubernetes Deployment via NodePort

### ISSUE THREAD

**Issue by lkuendig** [pgadmin-org/pgadmin4#6779]
**_Please note that security bugs or issues should be reported to security@pgadmin.org._**

**Describe the bug**

When I connect via JDBC connection string and Postrgres-driver or from an IntelliJ database-plugin it works. But adding the server from pgAdmin it doesn't. Error message:
Unable to connect to server: connection timeout expired

**To Reproduce**

https://stackoverflow.com/q/71335551/3417684

**Expected behavior**

Login works

**Error message**

Unable to connect to server: connection timeout expired

**Additional context**

- pgAdmin 4 - version 7.6
- postgres.conf: listen_addresses = '*'

**Comment by adityatoshniwal** [comment]
Hi @lkuendig,
This doesn't look like a pgAdmin issue. The issue here your minikube cluster is not accessible from the pgAdmin docker container. May be related to this - https://stackoverflow.com/questions/61686660/how-to-access-the-minikube-from-a-docker-container?

**Comment by lkuendig** [comment]
it's in the same cluster (k8s - not minikube) and the application (from different) host outside of the cluster can access postgres db. also I can access pgAdmin UI. therefore everything seems open from outside the cluster. just the connection between pgAdmin and postgres doesn't work

**Comment by adityatoshniwal** [comment]
Hi @lkuendig ,

Can you please share your k8 config yaml file?

**Comment by lkuendig** [comment]
- postgress deployment:
```
apiVersion: v1
kind: Deployment
metadata:
  labels:
    app: postgres
  name: postgres
  namespace: infrastructure
spec:
  replicas: 1
  selector:
    matchLabels:
      app: postgres
  strategy:
    type: Recreate
  template:
    metadata:
      labels:
        app/srv-default: "true"
        app: postgres
    spec:
      containers:
        - env:
            - name: POSTGRES_PASSWORD
              value: xxx
            - name: POSTGRES_USER
              value: xxx
            - name: POSTGRES_DB
              value: xxx
          image: postgres:latest
          name: postgres
          ports:
            - containerPort: 5432
              hostPort: 5432
              protocol: TCP
          resources: {}
          volumeMounts:
            - mountPath: /var/lib/postgresql/data
              name: postgres-claim0
      restartPolicy: Always
      volumes:
        - name: postgres-claim0
          persistentVolumeClaim:
            claimName: postgres-claim0
status: {}
```
- postgres service:
```
apiVersion: v1
kind: Service
metadata:
  labels:
    app: postgres
  name: postgres
  namespace: infrastructure
spec:
  type: NodePort
  ports:
    - name: "5432"
      port: 5432
      targetPort: 5432
      nodePort: 32543
  selector:
    app: postgres
status:
  loadBalancer: {}
```

- pgAdmin deployment:
```
apiVersion: apps/v1
kind: Deployment
metadata:
  labels:
    app: pg-admin
  name: pg-admin
  namespace: infrastructure
spec:
  replicas: 1
  selector:
    matchLabels:
      app: pg-admin
  strategy: {}
  template:
    metadata:
      labels:
        app/srv-default: "true"
        app: pg-admin
    spec:
      containers:
        - env:
            - name: PGADMIN_DEFAULT_EMAIL
              value: xxx@gmail.com
            - name: PGADMIN_DEFAULT_PASSWORD
              value: xxx
          image: dpage/pgadmin4
          name: pg-admin
          ports:
            - containerPort: 80
              hostPort: 5050
              protocol: TCP
          resources: {}
      restartPolicy: Always
status: {}
```
- pgAdmin service:
```
apiVersion: v1
kind: Service
metadata:
  labels:
    app: pg-admin
  name: pg-admin
  namespace: infrastructure
spec:
  type: NodePort
  ports:
    - name: "5050"
      port: 5050
      targetPort: 80
      nodePort: 32585
  selector:
    app: pg-admin
status:
  loadBalancer: {}
```

access pgAdmin-UI: http://192.168.1.180:32585 (ok)
access postgres via application: jdbc:postgresql://192.168.1.180:32543/xxx (ok)
access postgres via pgAdmin: 1) 192.168.1.180 - 2) 32543 - 3) xxx - 4) xxx - 5) xxx (NOK)

**Comment by yogeshmahajan-1903** [comment]
@lkuendig 

I am not able to reproduce the issue with above yaml configuration.(Of course I had to add 'PersistentVolumeClaim' for postgres). I have tested on minikune setup.
<img width="1792" alt="Screenshot 2023-09-26 at 4 47 44 PM" src="https://github.com/pgadmin-org/pgadmin4/assets/16016019/7c266d1c-b05b-4cc1-a4ad-ed4446913319">

Can you please postgres pod logs once?

Here is my yaml config - 
```

apiVersion: v1
# define a resource for volume chain
kind: PersistentVolumeClaim
metadata:
  # Name the volume chain
  name: postgres-claim0
spec:
  accessModes:
    # Allow ReadWrite to multiple pods
    - ReadWriteMany
  # PVC requesting resources
  resources:
    requests:
      # the PVC storage
      storage: 8Gi

---

apiVersion: apps/v1
kind: Deployment
metadata:
  labels:
    app: postgres
  name: postgres
spec:
  replicas: 1
  selector:
    matchLabels:
      app: postgres
  strategy:
    type: Recreate
  template:
    metadata:
      labels:
        app/srv-default: "true"
        app: postgres
    spec:
      containers:
        - env:
            - name: POSTGRES_PASSWORD
              value: xxx
            - name: POSTGRES_USER
              value: xxx
            - name: POSTGRES_DB
              value: xxx
          image: postgres:latest
          name: postgres
          ports:
            - containerPort: 5432
              hostPort: 5432
              protocol: TCP
          resources: {}
          volumeMounts:
            - mountPath: /var/lib/postgresql/data
              name: postgres-claim0
      restartPolicy: Always
      volumes:
        - name: postgres-claim0
          persistentVolumeClaim:
            claimName: postgres-claim0
status: {}

---
apiVersion: v1
kind: Service
metadata:
  labels:
    app: postgres
  name: postgres
spec:
  type: NodePort
  ports:
    - name: "5432"
      port: 5432
      targetPort: 5432
      nodePort: 32543
  selector:
    app: postgres
status:
  loadBalancer: {}
---

apiVersion: apps/v1
kind: Deployment
metadata:
  labels:
    app: pg-admin
  name: pg-admin
spec:
  replicas: 1
  selector:
    matchLabels:
      app: pg-admin
  strategy: {}
  template:
    metadata:
      labels:
        app/srv-default: "true"
        app: pg-admin
    spec:
      containers:
        - env:
            - name: PGADMIN_DEFAULT_EMAIL
              value: abc@gmail.com
            - name: PGADMIN_DEFAULT_PASSWORD
              value: xxxx
          image: dpage/pgadmin4
          name: pg-admin
          ports:
            - containerPort: 80
              hostPort: 5050
              protocol: TCP
          resources: {}
      restartPolicy: Always
status: {}

---
apiVersion: v1
kind: Service
metadata:
  labels:
    app: pg-admin
  name: pg-admin
spec:
  type: NodePort
  ports:
    - name: "5050"
      port: 5050
      targetPort: 80
      nodePort: 32585
  selector:
    app: pg-admin
status:
  loadBalancer: {}
```

**Comment by lkuendig** [comment]
thank you for reproducing! postgres-log is empty

I'm connecting with pgAdmin from a different host than my cluster/node ist running on. Are you still able to connect in this scenario? Then it's somewhere a problem on my side :(

**Comment by lkuendig** [comment]
My app runs without a problem as well I'm able to view/edit everything from my IntelliJ database browser. I guess pgAdmin connects to the nodePort (which is possible) but afterwards switches to the container-port which then is not accessible from outside the cluster (where pgAdmin is used for connecting). could that be? the request that fails:
`2023-10-15T15:04:32.340265050Z ::ffff:10.42.0.1 - - [15/Oct/2023:15:04:32 +0000] "POST /browser/server/obj/1/ HTTP/1.1" 401 121 "http://192.168.1.180:32585/browser/" "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/116.0.0.0 Safari/537.36"`

**Comment by yogeshmahajan-1903** [comment]
@lkuendig 
Can you please share error screen shot and pgadmin pod logs? I was trying on minikune, hence I had same host where Kubernetes cluster present. If you are able to login to pgadmin , then pgadmin is working with correct port. Is the postgres service accessible from same host where you have launched pgAdmin 4?

**Comment by lkuendig** [comment]
Here is my setup. Everything in green works. Except the one in red. And the answer for your question: yes IntelliJ connection works from my laptop (node-port). and also container-port works (app which is deployed in kubernetes)
![PgAdmin2](https://github.com/pgadmin-org/pgadmin4/assets/56652114/fce15c7c-9d6d-40ca-85c1-94e670476c67)

Here are all the logs (fresh/new pod):
[pgAdmin_log.txt](https://github.com/pgadmin-org/pgadmin4/files/13244281/pgAdmin_log.txt)

And here is the error message:
![Screenshot from 2023-11-02 20-58-17](https://github.com/pgadmin-org/pgadmin4/assets/56652114/ef1a65b2-f950-459e-a678-6334a03b2e1a)

Thank you for taking a look!

**Comment by yogeshmahajan-1903** [comment]
@lkuendig 
I am able to reproduce the issue where I have deployed pgAdmin 4 and PostgreSQL on AKS(Azure). To resolve the problem, I had to add IP of all nodes in Kubernetes cluster to AKS cluster security group. Basically, allow cluster nodes to accept request from external ips of the nodes, as incoming ip will be not machine ip but pgadmin pod ip.


<img width="1790" alt="Screenshot 2023-12-05 at 6 31 35 PM" src="https://github.com/pgadmin-org/pgadmin4/assets/16016019/387ffe60-5a0b-4669-a1e9-3180e92c6804">
